### PR TITLE
feat(Container): list collection can be a List of items or a Map

### PR DESCRIPTION
**Pull request by jsomsanith**
**Please check if the PR fulfills these requirements**

- [x] The commit(s) message(s) follows our
  [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md#commit-message-format)
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What is the current behavior?** (You can also link to an open issue here)
The collections can only be a `List` to be displayed by the List container.
But we need metadata sometimes, like for pagination, or infinite scroll, where the server can send back extra info. Those metadata should accompany the collection items.

**What is the new behavior?**
A CMF collection can be 
* a List containing the items
* a Map containing the items under `item` property, and extra info. In that case, the List container extract the `pagination` property to add them in the toolbar props.

This solves #176 

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://180.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://180.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://180.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://180.talend.surge.sh)

**Comment by talend-glorieux** [comment]
Could you please add a story of the container? I'd like to see what it looks like in action.

File: packages/containers/src/List/List.connect.js [hunk 1]
```diff
@@ -21,18 +30,16 @@ export function mapDispatchToProps(dispatch, ownProps) {
...
-	})(state);
+	};
 
-	if (items) {
-		items = items.toJS();
-	} else {
-		items = [];
+	props.items = getItems(state, config);
+	if (props.state && props.state.has('toolbar')) {
+		const dynamicPagination = configureGetPagination(state, config);
```
Comment (talend-glorieux) [inline packages/containers/src/List/List.connect.js]: I don't think that assignation is required.

File: packages/containers/src/List/selector.js [hunk 2]
```diff
@@ -1,7 +1,7 @@
 import { createSelector } from 'reselect';
 import { Map, List } from 'immutable';
 
-export function contains(ob, query) {
+function contains(ob, query) {
```
Comment (talend-glorieux) [inline packages/containers/src/List/selector.js]: What is `ob`?

Comment (jsomsanith) [inline packages/containers/src/List/selector.js]: I haven't touched that code except the export. Looking at the code it seems that this is a list item, so any object. I'll rename it to `listItem`

File: packages/containers/src/List/List.test.js [hunk 3]
```diff
@@ -140,39 +148,123 @@ describe('Container List', () => {
...
 		expect(Connected.WrappedComponent).toBe(Container);
 	});
-	it('should map state to props', () => {
+
+	it('should map items to props from collection List', () => {
+		// given
 		const state = {
 			cmf: {
-				components: new Map({
+				components: Immutable.fromJS({
```
Comment (acateland) [inline packages/containers/src/List/List.test.js]: fromJS can be imported alone

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://180.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://180.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://180.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://180.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://180.talend.surge.sh)

**Comment by jmfrancois** [comment]
:octocat: [Demo is available here](http://180.talend.surge.sh)

**Pull request merged**